client/network/Cargo.toml

```diff
@@ -16,6 +16,7 @@ targets = ["x86_64-unknown-linux-gnu"]
 prost-build = "0.6.1"
 
 [dependencies]
+async-std = { version = "1.6.2", features = ["unstable"] }
```

Comment:
No, it is for `Convar`. It is unfortunate that we have to depend on the `unstable` feature, but I couldn't find any crate other than `async-std` that provides an asynchronous `Condvar`.

Reply:
If this is something we want to avoid, one can use a channel as a condvar.

```rust
use futures::channel::mpsc::channel;

fn main() {
    let (tx, rx) = channel(0);
    
    let producer = async move {
        // Produce something and put it to place X.
        
        // Signal the consumer that that something can be optained.
        tx.send(()).await;
    }
    
    let consumer = async move {
        // Wait for the signal. Maybe include a timeout like you do today with 10 secs.
        let _ = rx.next().await;
        
        // Access place X.
    }
}
```

Problem is that this can't be used in the `Drop` implementations and thus cleanup would depend on the 10 sec timeout.

Reply:
> Problem is that this can't be used in the Drop implementations [...]

It does not always need to. When the `Sender` is dropped, the receiver will notice and the task can terminate. Instead of `let _ = rx.next().await;` one would write `if rx.next().await.is_some() { ... }`. In the Drop impl of QueueGuard one can use `Sender::try_send`.

Reply:
I find it a bit weird to use a channel, which involves an additional `Arc`, `Mutex` and `Vec`, just to wake up a task, rather than a `Waker`.

I have now also tried using a `Waker`, but the implementation is considerably more tricky because of potential race conditions and having to introduce manual polling within an `async` function and having to implement your own `Waker`.

I don't really understand why we don't want the `Condvar` solution, as a `Condvar` is exactly the tool that is designed for this specific job.